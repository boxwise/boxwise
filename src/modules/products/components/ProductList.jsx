import React, { useEffect } from "react";
import { withRouter, Link } from "react-router-dom";
import { withStyles } from "@material-ui/core";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import Tooltip from "@material-ui/core/Tooltip";
import TableRow from "@material-ui/core/TableRow";
import Snackbar from "@material-ui/core/Snackbar";
import Fab from "@material-ui/core/Fab";
import IconButton from "@material-ui/core/IconButton";
import EditIcon from "@material-ui/icons/Edit";
import AddIcon from "@material-ui/icons/Add";
import DeleteIcon from "@material-ui/icons/Delete";

import Progress from "components/Progress";
import ConfirmationButton from "components/ConfirmationButton";

const styles = theme => ({
  fab: {
    position: "absolute",
    bottom: theme.spacing.unit * 2,
    right: theme.spacing.unit * 2
  }
});

const ProductList = ({
  products: { loading, data },
  location,
  history,
  deleteProduct,
  getAllProducts,
  classes
}) => {
  const hideMessage = () => {
    history.replace({ ...location.pathname, state: undefined });
  };
  useEffect(() => {
    // TODO: instead of loading every time this page loads, we should find a way to
    // subscribe to products in firebase and keep them locally
    getAllProducts();
  }, [getAllProducts]);

  return (
    <React.Fragment>
      {loading && <Progress />}
      <Table>
        <TableHead>
          <TableRow>
            <TableCell padding="dense">Category</TableCell>
            <TableCell padding="dense">Name</TableCell>
            <TableCell padding="dense" />
          </TableRow>
        </TableHead>
        <TableBody data-testid="productsTableBody">
          {data.map(product => (
            <TableRow key={product.id}>
              <TableCell padding="dense">{product.category}</TableCell>
              <TableCell padding="dense" data-testid="productNameCell">
                {product.name}
              </TableCell>
              <TableCell padding="dense">
                <IconButton
                  component={Link}
                  to={`/products/edit/${product.id}`}
                  aria-label="Edit"
                  data-testid="editProductButton"
                >
                  <EditIcon />
                </IconButton>
                <ConfirmationButton
                  confirmationAction="Delete"
                  confirmationText="Are you sure you want to delete this product?"
                  onConfirm={() => deleteProduct(product.id)}
                  dataTestId="deleteProduct"
                >
                  <DeleteIcon />
                </ConfirmationButton>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
      <Tooltip title="Add product" aria-label="Add">
        <Fab
          color="secondary"
          className={classes.fab}
          data-testid="addProductButton"
          component={Link}
          to="/products/add"
        >
          <AddIcon />
        </Fab>
      </Tooltip>
      <Snackbar
        anchorOrigin={{
          vertical: "bottom",
          horizontal: "left"
        }}
        onClose={hideMessage}
        open={location.state && location.state.message !== undefined}
        autoHideDuration={4000}
        message={location.state && location.state.message}
      />
    </React.Fragment>
  );
};

export default withRouter(withStyles(styles)(ProductList));
