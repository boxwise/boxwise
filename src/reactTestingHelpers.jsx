import React from "react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { render, fireEvent } from "react-testing-library";
import { createMuiTheme, MuiThemeProvider } from "@material-ui/core/styles";
import thunk from "redux-thunk";
import { createStore, applyMiddleware } from "redux";

export { fireEvent, wait } from "react-testing-library";

// following suggestions in https://testing-library.com/docs/react-testing-library/setup#custom-render

export const typeText = (element, value) => {
  fireEvent.change(element, { target: { value } });
};

const theme = createMuiTheme({
  typography: {
    useNextVariants: true
  }
});

export function customRender(
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
    initialState = {},
    reducer = state => state
  } = {}
) {
  const store = createStore(reducer, initialState, applyMiddleware(thunk));
  return render(
    <MuiThemeProvider theme={theme}>
      <Provider store={store}>
        <Router history={history}>{ui}</Router>
      </Provider>
    </MuiThemeProvider>
  );
}

export { customRender as render };
