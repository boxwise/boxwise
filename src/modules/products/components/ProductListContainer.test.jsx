import React from "react";

import { render, wait } from "reactTestingHelpers";

import ProductList from "./ProductListContainer";

jest.mock("modules/products/api", () => ({
  getAllProducts: async () => [{ id: "abc", name: "hello" }]
}));

describe("ProductsPageContainer", () => {
  it("renders correctly with a simple list", async () => {
    const { container, store } = render(<ProductList />);

    expect(container).toMatchSnapshot("loading data");
    expect(JSON.stringify(store.getState(), null, 2)).toMatchSnapshot(
      "store loading data"
    );

    await wait();

    expect(container).toMatchSnapshot("loaded data");
    expect(JSON.stringify(store.getState(), null, 2)).toMatchSnapshot(
      "store loaded data"
    );
  });
  // TODO: more tests, yes I know...
});
