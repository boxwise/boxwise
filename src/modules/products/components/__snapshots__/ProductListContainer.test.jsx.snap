// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsPageContainer renders correctly with a simple list: loaded data 1`] = `
<div>
  <table
    class="MuiTable-root-2"
  >
    <thead
      class="MuiTableHead-root-3"
    >
      <tr
        class="MuiTableRow-root-4 MuiTableRow-head-7"
      >
        <th
          class="MuiTableCell-root-9 MuiTableCell-head-10 MuiTableCell-paddingDense-14"
          scope="col"
        >
          Category
        </th>
        <th
          class="MuiTableCell-root-9 MuiTableCell-head-10 MuiTableCell-paddingDense-14"
          scope="col"
        >
          Name
        </th>
        <th
          class="MuiTableCell-root-9 MuiTableCell-head-10 MuiTableCell-paddingDense-14"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root-21"
      data-testid="productsTableBody"
    >
      <tr
        class="MuiTableRow-root-4"
      >
        <td
          class="MuiTableCell-root-9 MuiTableCell-body-11 MuiTableCell-paddingDense-14"
        />
        <td
          class="MuiTableCell-root-9 MuiTableCell-body-11 MuiTableCell-paddingDense-14"
          data-testid="productNameCell"
        >
          hello
        </td>
        <td
          class="MuiTableCell-root-9 MuiTableCell-body-11 MuiTableCell-paddingDense-14"
        >
          <a
            aria-label="Edit"
            class="MuiButtonBase-root-40 MuiIconButton-root-66"
            data-testid="editProductButton"
            href="/products/edit/abc"
            role="button"
            tabindex="0"
          >
            <span
              class="MuiIconButton-label-71"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-43"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-59"
            />
          </a>
          <button
            aria-label="Delete"
            class="MuiButtonBase-root-40 MuiIconButton-root-66"
            data-testid="deleteProductButton"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label-71"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-43"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-59"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <a
    aria-label="Add"
    class="MuiButtonBase-root-40 MuiFab-root-30 MuiFab-secondary-33 ProductList-fab-1"
    data-testid="addProductButton"
    href="/products/add"
    role="button"
    tabindex="0"
    title="Add product"
  >
    <span
      class="MuiFab-label-31"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-43"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root-59"
    />
  </a>
</div>
`;

exports[`ProductsPageContainer renders correctly with a simple list: loading data 1`] = `
<div>
  <img
    alt="progress"
    class=""
    src="/images/loading-boxes.gif"
    style="height: 40px; width: 40px;"
  />
  <table
    class="MuiTable-root-2"
  >
    <thead
      class="MuiTableHead-root-3"
    >
      <tr
        class="MuiTableRow-root-4 MuiTableRow-head-7"
      >
        <th
          class="MuiTableCell-root-9 MuiTableCell-head-10 MuiTableCell-paddingDense-14"
          scope="col"
        >
          Category
        </th>
        <th
          class="MuiTableCell-root-9 MuiTableCell-head-10 MuiTableCell-paddingDense-14"
          scope="col"
        >
          Name
        </th>
        <th
          class="MuiTableCell-root-9 MuiTableCell-head-10 MuiTableCell-paddingDense-14"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root-21"
      data-testid="productsTableBody"
    />
  </table>
  <a
    aria-label="Add"
    class="MuiButtonBase-root-40 MuiFab-root-30 MuiFab-secondary-33 ProductList-fab-1"
    data-testid="addProductButton"
    href="/products/add"
    role="button"
    tabindex="0"
    title="Add product"
  >
    <span
      class="MuiFab-label-31"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-43"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </span>
    <span
      class="MuiTouchRipple-root-59"
    />
  </a>
</div>
`;

exports[`ProductsPageContainer renders correctly with a simple list: store loaded data 1`] = `
"{
  \\"products\\": {
    \\"byId\\": {
      \\"abc\\": {
        \\"id\\": \\"abc\\",
        \\"name\\": \\"hello\\"
      }
    },
    \\"allIds\\": [
      \\"abc\\"
    ],
    \\"loading\\": false
  },
  \\"user\\": {
    \\"hasInitialized\\": false,
    \\"loading\\": false,
    \\"isUpdating\\": false,
    \\"hasTriggeredReset\\": false
  },
  \\"boxes\\": {
    \\"byId\\": {},
    \\"allIds\\": [],
    \\"loading\\": false
  }
}"
`;

exports[`ProductsPageContainer renders correctly with a simple list: store loading data 1`] = `
"{
  \\"products\\": {
    \\"byId\\": {},
    \\"allIds\\": [],
    \\"loading\\": true
  },
  \\"user\\": {
    \\"hasInitialized\\": false,
    \\"loading\\": false,
    \\"isUpdating\\": false,
    \\"hasTriggeredReset\\": false
  },
  \\"boxes\\": {
    \\"byId\\": {},
    \\"allIds\\": [],
    \\"loading\\": false
  }
}"
`;
